### PR TITLE
ISSUE-18 Added entropy bits estimate value based on NIST-SP-800-53-1

### PR DESCRIPTION
**Pull request by ememisya**
-Implemented entropy estimate as defined in http://csrc.nist.gov/publications/nistpubs/800-63-1/SP-800-63-1.pdf
-Bumped minor version due to structural changes I've concluded to be more efficient.
-Refactoring

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -7,7 +7,7 @@
   <groupId>org.passay</groupId>
   <artifactId>passay</artifactId>
   <packaging>jar</packaging>
-  <version>1.2.0-SNAPSHOT</version>
+  <version>1.3.0-SNAPSHOT</version>
```
Comment (dfish3r) [inline pom.xml]: 1.2.0 hasn't been released. No need to bump the minor version.

File: src/main/java/org/passay/AbstractDictionaryRule.java [hunk 2]
```diff
@@ -76,18 +76,20 @@ public RuleResult validate(final PasswordData passwordData)
...
+        result.getDetails().add(new RuleResultDetail(ERROR_CODE, createRuleResultDetailParameters(matchingWord)));
+      }
+      if (matchBackwards) {
+        text = new StringBuilder(passwordData.getPassword()).reverse().toString();
+        matchingWord = doWordSearch(text);
+        if (matchingWord != null) {
+          result.setValid(false);
+          result.getDetails().add(
+            new RuleResultDetail(ERROR_CODE_REVERSED, createRuleResultDetailParameters(matchingWord)));
+        }
```
Comment (dfish3r) [inline src/main/java/org/passay/AbstractDictionaryRule.java]: What's the purpose of this change? Seems unrelated to the task at hand.

File: src/main/java/org/passay/AllowedCharacterRule.java [hunk 3]
```diff
@@ -11,14 +11,14 @@
  *
  * @author  Middleware Services
  */
-public class AllowedCharacterRule implements Rule
+public class AllowedCharacterRule implements CharacterData, Rule
```
Comment (dfish3r) [inline src/main/java/org/passay/AllowedCharacterRule.java]: That interface wasn't meant for rules, what problem are you solving with it?

File: src/main/java/org/passay/CharacterRule.java [hunk 4]
```diff
@@ -48,7 +48,7 @@ public CharacterRule(final CharacterData data, final int num)
    *
    * @param  n  number of characters to require where n &gt; 0
    */
-  public void setNumberOfCharacters(final int n)
+  private void setNumberOfCharacters(final int n)
```
Comment (dfish3r) [inline src/main/java/org/passay/CharacterRule.java]: Why change this to private? Most of the provided rules are not immutable.

File: src/main/java/org/passay/IllegalCharacterRule.java [hunk 5]
```diff
@@ -11,7 +11,7 @@
  *
  * @author  Middleware Services
  */
-public class IllegalCharacterRule implements Rule
+public final class IllegalCharacterRule implements Rule
```
Comment (dfish3r) [inline src/main/java/org/passay/IllegalCharacterRule.java]: The provided rules are intended for extension, why change this?

File: src/main/java/org/passay/PasswordValidator.java [hunk 6]
```diff
@@ -55,14 +59,114 @@ public PasswordValidator(final MessageResolver resolver, final List<Rule> rules)
...
+   *
+   * @param  passwordData  to validate
+   * @throws org.passay.EntropyException Thrown when passwordData {@link PasswordOrigin} is not implemented
+   * by this method or if the required {@link CharacterRule} instances are unavailable in the passwordRules of
+   * this validator instance.
+   *
+   * @see <a href="http://csrc.nist.gov/publications/nistpubs/800-63-1/SP-800-63-1.pdf">PDF Publication</a>
+   * @return  rule result
+   */
+  public Double getEntropyBits(final PasswordData passwordData)
```
Comment (dfish3r) [inline src/main/java/org/passay/PasswordValidator.java]: This feels like it belongs in a different class. EntropyCalculator#calculateEntropy perhaps.

**Comment by dfish3r** [comment]
Overall this change looks good. Just needs some tweaks to the API.

File: pom.xml (hunk 1 above)
Comment (ememisya) [inline pom.xml]: That makes sense, I should have checked that first :)

File: src/main/java/org/passay/AbstractDictionaryRule.java (hunk 2 above)
Comment (ememisya) [inline src/main/java/org/passay/AbstractDictionaryRule.java]: Ah, that can be left out as it is indeed unrelated to the issue.  The purpose was that since we are checking for empty string on the backwards search, we should do so on the forward search as well.

File: src/main/java/org/passay/AllowedCharacterRule.java (hunk 3 above)
Comment (ememisya) [inline src/main/java/org/passay/AllowedCharacterRule.java]: The problem solves is the ability to get a CharSequence out of similar rules such as AllowedCharacterRule and CharacterRule, otherwise I needed to use getAllowedCharacters and getValidCharacters which in essence does the same thing.  I referred to this under the refactoring message.

File: src/main/java/org/passay/CharacterRule.java (hunk 4 above)
Comment (ememisya) [inline src/main/java/org/passay/CharacterRule.java]: My thought pattern behind this was that since we have the setter under the constructor, and it is public yet not final, extending this class might yield to unexpected behavior.  This was also the reason why I declared some public Rules as final, solely due to having a non final public method under the constructor.

File: src/main/java/org/passay/IllegalCharacterRule.java (hunk 5 above)
Comment (ememisya) [inline src/main/java/org/passay/IllegalCharacterRule.java]: Err, as mentioned above, this was also due to public non-final setter methods being used under the constructor.  I suppose I could also make the setters final instead.  Or have a private final both the constructor and the setters refer to.

File: src/main/java/org/passay/PasswordValidator.java (hunk 6 above)
Comment (ememisya) [inline src/main/java/org/passay/PasswordValidator.java]: Well, I can totally see that.  I wanted to keep the entropy calculation native to the validator solely from the library user's perspective, and the fact that it can be considered a password utility (all static).  Feel free to merge separately, I personally wanted to put all the Rule implementations and the interface under a rule package, but thought that might be a major version.
